c++/src/kj/common.h

```diff
@@ -1748,6 +1749,18 @@ class ArrayPtr: public DisallowConstCopyIfNotConst<T> {
     KJ_IREQUIRE(start <= end && end <= size_, "Out-of-bounds ArrayPtr::slice().");
     return ArrayPtr(ptr + start, end - start);
   }
+  inline bool startsWith(const ArrayPtr<const T>& other) const {
+    return other.size_ <= size_ && slice(0, other.size_) == other;
+  }
+
+  inline Maybe<size_t> findFirst(char c) const {
```

Comment:
This only makes sense if `T` is `char`. If the array is some other type, this is a weird operation that would likely confuse people. For example, if you had an `ArrayPtr<int>` and tried to call `array.findFirst(1)`, it would surprisingly not find the first element with value 1, but would instead find the first individual byte with value 1.

If you want to add `findFirst` I think its parameter needs to be type `const T&`, and you'd have to implement it using a generic `for` loop that tests each element for equality with `==`. You could then implement template explicit specializations for `char` and `byte` that use `memchr`.